### ISSUE TITLE
Equality fools the totality checker, proving Void

### ISSUE THREAD

**Issue by gelisam** [idris-lang/Idris-dev#1833]
While the totality checker correctly rejects the following negative occurrence of `Con1`,

    data Con1 : Type where
      con1 : (Con1 -> Void) -> Con1

    runCon1 : Con1 -> (Con1 -> Void)
    runCon1 (con1 f) = f

the totality checker accepts the following equivalent definition, which uses an equality constraint to hide the fact that the negatively-occurring type must be `Con2`.

    data Con2 : Type where
      con2 : (a : Type) -> Con2 = a -> (a -> Void) -> Con2

    runCon2 : Con2 -> (Con2 -> Void)
    runCon2 (con2 _ Refl f) = f

See https://gist.github.com/gelisam/c4df02dc451d7ccf9307 for details on how this can be used to derive `Void`.

Tested on HEAD, currently 0.9.15.1-git:b0257ec.

**Comment by ahmadsalim** [comment]
Did you try using `%default total` at the start of the file or have a total annotation before the function, e.g.:

```
total
runCon2 : Con2 -> (Con2 -> Void)
runCon2 (con2 _ Refl f) = f
```
?

As far as I recall, Idris does not enforce totality by default; so this might be why the issue is not caught.

**Comment by edwinb** [comment]
Whether or not there is a `%default total` annotation, Idris *always* checks for totality.

Most likely the checker is simply not looking under the '=' constructor so not spotting the negative occurrence. However, I haven't looked either (busy with other stuff at the moment, sorry...)

**Comment by ahmadsalim** [comment]
Unlike your first example, it is actually impossible to prove `Con2 = a -> (a -> Void)` in the second example, and so it would not be possible to call `runCon2` with anything that returns void.

**Comment by gelisam** [comment]
I must respectfully point out that my definition of `Con2` is not

    data Con2 : Type where
      con2 : (a : Type) -> (Con2 = a -> (a -> Void)) -> Con2

but rather

    data Con2 : Type where
      con2 : (a : Type) -> Con2 = a -> (a -> Void) -> Con2

That is, in order to produce a value of type `Con2`, we don't need to produce a value of type `Con2 = a -> (a -> Void) `, but rather, we need to produce a value of type `Con2 = a` and a value of type `a -> Void`. For example, we can use `Refl` and a value of type `Con2 -> Void`.

I would also like to point out that while I have only included the code for a value of type `Con2 -> (Con2 -> Void)` in this comment, the gist I linked to in my first comment does derive a value of type `Void`, and therefore this is a real bug, not an hypothetical.
